### ISSUE TITLE
Location of *.parts files

### ISSUE THREAD

**Issue by SpL1Ne** [arvidn/libtorrent#7194]
https://github.com/qbittorrent/qBittorrent/issues/13205
As I can see in [this comment](https://github.com/qbittorrent/qBittorrent/issues/13205#issuecomment-665686667) there is no way to define folder for parts file. Also I searched and didn't find issue about this question.
So, is there in plans to provide possibility to change location of *.parts file (instead near downloaded torrent files)?

libtorrent version (or branch):
qBittorrent 4.4.5

platform/architecture:
Windows 10x64 21H1

**Comment by master255** [comment]
What is it for? A cosmetic effect? Or functional?

**Comment by SpL1Ne** [comment]
Well, users literally has some hidden files inside download folder. I'm usually move it whole and capture batch of 100-1000mb file that don't need to. And regular user doesn't even know about them.

So yeah definitely functional. Keep regular files at one location and qbittorrent files at other and don't mix them like don't mix temp files and user's files in any other program.

**Comment by master255** [comment]
@SpL1Ne 

> Well, users literally has some hidden files inside download folder. I'm usually move it whole and capture batch of 100-1000mb file that don't need to. And regular user doesn't even know about them.

I don't understand what you're writing about. Can you explain it in more detail?

**Comment by SpL1Ne** [comment]
@master255 
1. Let's say you have like 5 torrents with same theme\subject. For example, pictures, because they have more parts to complete pieces, if you don't download whole torrent.
2. Since all of them have same theme, you create folder to store them together. In my case fore example `Pictures` folder inside `Download` folder.
3. You load torrent file into qbittorrent. Select let's say 1/5 of total files. Because you don't need whole torrent.
4. Do that with other 4 torrents.
5. Set for all of them location as `Pictures` folder inside `Download`
6. Wait torrents to complete.
7. Now, you have your `Pictures` folder with torrents what you want.
8. You don't store your media inside `Download` folder, so you select `Pictures` folder ctrl-x and ctrl-v inside `D:\Media`.
9. Seems like everything good, right?
10. Wrong.
![2023-03-19_00-50-45](https://user-images.githubusercontent.com/8836461/226144504-b0750726-f8ab-48e5-b5a1-479ee893ab72.png)
11. You downloaded 1 Gb from 5Gb torrent and also you have `*.part` files INSIDE `Pictures` folder that you will move after all torrents complete. If you don't know about them, you just move 5*1 Gb folders and PLUS ~100-200 Mb overhead for each folder. So inside your `Media` folder now 1Gb of useless files that you don't know about.

Note: qBittorrent v4.4.5, don't know if it's latest version or if devs do something with that situation.

**Comment by master255** [comment]
@SpL1Ne So what's the problem?

**Comment by SpL1Ne** [comment]
OMG, really? You don't see any problem, do you?

That I, as user, have useless overhead that take space and if I didnt know about it I'd have a lot of useless garbage files with files that I needed. I can copy files to USB flash drive to move them and I also would move garbage together and wouldn't know about it.
Yeah, agree, no problems here. /sarcasm

**Comment by master255** [comment]
@SpL1Ne Ok. When I download torrents I don't have any extra files left on my disk, after the full download. Why do you have them left? 
If you don't need these files, why not delete them?

**Comment by SpL1Ne** [comment]
> after the full download

Are you stupid or what? Quote from my message above:

> Select let's say 1/5 of total files. Because you don't need whole torrent.

/facepalm

> If you don't need these files, why not delete them?

BECAUSE THESE FILES ARE HIDDEN, holy. If you don't know about them, then you CAN'T delete them. This is ME, I know about them. And 99% percent of other users don't.

I won't answer to your messages any more since you definitely don't understand words and don't read what I write here.

**Comment by soredake** [comment]
> I also don't get the problem. When you delete a task from the client, corresponding `.parts` file is also deleted along with it automatically.

Not always, sometimes they are not deleted.

**Comment by stalkerok** [comment]
@SpL1Ne, your problem is that you are manually transferring files instead of using the move feature in qbt. Then delete the task without deleting the files, the .parts file will be deleted automatically.

**Comment by SpL1Ne** [comment]
@stalkerok 
> your problem is that you are manually transferring files instead of using the move feature in qbt. 

My problem? LMAO. Didn't see any warnings that app will not tolerate external manipulation with files. I know, right? Where do you ever see a user manually moving downloaded files? Nonsense.
This is app problem, that it conflicts with external manipulation, not mine.

**Comment by soredake** [comment]
Not stale....

**Comment by cheater** [comment]
can this become a thing? it's pretty annoying. I can't find anything at all in my download directory:

![qbittorrent parts files](https://github.com/arvidn/libtorrent/assets/195390/723bed57-c4c8-4a12-bb31-fe3665749f85)

**Comment by HanabishiRecca** [comment]
> I can't find anything at all in my download directory:

As a workaround, you can temporary disable Explorer "show hidden files" option.

Parts were made hidden with user not seeing them in mind. But many advanced users have hidden files enabled, so yeah.

**Comment by cheater** [comment]
while this is *a* workaround, personally i don't really like it and honestly, guys, it's not so deep, this should be a thing...

**Comment by glassez** [comment]
Wouldn't it make sense to store .parts files in TEMP folder?

**Comment by cheater** [comment]
it will most likely be a different file system, and especially someone's OS file system: added wear means failure of OS SSD = people get angry; plus filling up the OS SSD when on many computers it's tiny (sometimes 128 or 256 GB) is bad.

**Comment by SpL1Ne** [comment]
> Wouldn't it make sense to store .parts files in TEMP folder?

And also because parts files are not temp files. You need them if you want to seed torrent, but you don't want to download all files from that torrent. That's because how bittorrent protocol works.

**Comment by HanabishiRecca** [comment]
> Wouldn't it make sense to store .parts files in TEMP folder?

No, it tend to be cleaned up by users, Windows Disk Cleanup and other utilities. It is called temporary for a reason. Also will be lost on system switch/reinstall etc.
Besides, putting the actual data on the system drive could be really undesirable behavior, as already mentioned above.

Pretty much the same applies for other operating systems.

---

Btw, placing part files inside the torrent subfolder is actually achievable, in a somewhat hacky way though.
If you set content layout to `Don't create subfolder` and make it as the target download location manually.

![](https://i.ibb.co/WxvdNZm/image.png)

![](https://i.ibb.co/brgcF8K/image.png)

**Comment by glassez** [comment]
> So instead of `<download_dir>/.<info_hash>.parts` it will be `<download_dir>/.parts/.<info_hash>.parts`.
> Naming is arguable of course.

IMO, it's a good compromise.

**Comment by cheater** [comment]
> So instead of `<download_dir>/.<info_hash>.parts` it will be `<download_dir>/.parts/.<info_hash>.parts`. Naming is arguable of course.

yes, that is perfect, and an easy change, please don't wait and put it in :) no deliberation is needed.

one thing is necessary though: moving parts files for existing torrents to new location. idk how torrent clients handle that. having to delete and re-dl all the data in parts files is not an option, there must be a way to retain the data. it should all probably happen automatically the first time you start a torrent client with new libtorrent that supports a .parts subdir.

**Comment by SpL1Ne** [comment]
> IMO, it's a good compromise.

I would like to set one folder for all parts files and keep download folder clean and apart from "working" files so to say.

**Comment by HanabishiRecca** [comment]
> it should all probably happen automatically the first time you start a torrent client with new libtorrent that supports a .parts subdir.

Pretty much impossble at this point. The location of files is handled by libtorrent. Changing the harcoded path simply changes where it searches for them. There is no code to handle such migration.
That's why this is not so "easy change" as you say. Simply change the path - easy. Properly migrate all users - not easy.

**Comment by cheater** [comment]
it's not so hard.

1. check old path. if exists, move to new path
2. continue with new path

make sure to add a comment to the check in (1) that's like "remove this check in 2029" or w/e so it doesn't check the old path forever...

**Comment by cheater** [comment]
> > IMO, it's a good compromise.
> 
> I would like to set one folder for all parts files and keep download folder clean and apart from "working" files so to say.

let's get the `.parts` subdir done first, what you're talking about is quite a bit more advanced. besides, you'd need at least one parts folder per file system, and it's difficult to figure out how to do this... either case open a new issue so discussion can happen here, not interrupted by whatever we're talking about in here.

**Comment by cheater** [comment]
Are we on course to forgetting this for another 4 years? It seems like people got hung up on how to handle old parts files, so here's the way to do this:

1. now there are two possible paths for parts files: old and new. new includes the .parts subdir, old doesn't
2. check if the parts file for our given torrent is in the old path. if it is, move it to the new path
3. continue doing whatever you'd do by using the new file path

How is moving files done elsewhere in libtorrent? does it directly call a posix syscall, or is there some sort of compatibility layer?

**Comment by cheater** [comment]
if both paths exist:

1. if one of the files is zero bytes, ignore it and use the other one
2. if both files are non-zero, use the larger one
3. if both files are the same size, use the newer one
4. in any case delete the other file

**Comment by glassez** [comment]
> Are we on course to forgetting this for another 4 years?

@cheater 
Would you like to do it yourself? You're a "dev", as you've mentioned many times.

**Comment by cheater** [comment]
This requires effort from multiple parties. Doing solo dev by someone who isn't familiar with the code base will only result in breaking it. Keeping track of issues so they don't get forgotten about is as important as any other piece of effort. Would you like to do this? Maybe rather than respond with snark, recognize that it's important to not leave stuff on the backburner forever.

**Comment by cheater** [comment]
> > Are we on course to forgetting this for another 4 years?
> 
> @cheater Would you like to do it yourself? You're a "dev", as you've mentioned many times.

Meanwhile, can we focus on the technical issue at hand, and answer the question?

> How is moving files done elsewhere in libtorrent? does it directly call a posix syscall, or is there some sort of compatibility layer?

**Comment by cheater** [comment]
@soredake if you have anything to add that will move this issue forward, feel free. otherwise take your passive-aggressive thumbs down reactions and shove em

**Comment by SpL1Ne** [comment]
Can you stop flooding messages? If you have something to say, say that in ONE message, ffs.
Try to learn how to use "edit" if you have more thoughts to add to previous message.

**Comment by cheater** [comment]
Nope, sorry.

**Comment by arvidn** [comment]
The main reason the path to the part file is hard coded/well defined relative to the download path is to be able to, automatically and reliably, pick it up when adding a torrent to the session with the intention to seed existing files.

Perhaps this is not the most important use case, most of the time you probably have resume data for the torrent, and it could store the path to the part file.

Even just changing the hard-coded path wouldn't be entirely trivial, as it would need to look for the part file in two different places. The upgrade logic along might dominate the effort, so perhaps taking it all the way and making it configurable in settings makes more sense.

**Comment by glassez** [comment]
@arvidn 
"Part file" has a vague position in `libtorrent` architecture. On the one hand, it looks like an implementation detail of a disk IO subsystem that is not defined in any way in its interface. On the other hand, it still leaks a little into the interface (in the form of `delete_partfile` flag).

**Comment by glassez** [comment]
@arvidn 
How would you expect "part file" to be customizable?
It seems to me that this could be done through `add_torrent_params`. In addition, this should entail adding a new method to change this setting (moving the "part file" to another location). Even if it were generally approved, you are unlikely to allow such changes to the current (at the time when these changes are proposed) branch due to API/ABI incompatibility.

**Comment by arvidn** [comment]
In an ideal world, per-disk subsystem settings would be set on the concrete disk subsystem object directly. That way, further updates and tweaks to disk settings can be done without affecting other disk subsystems nor as likely to affect ABI.

**Comment by cheater** [comment]
does a feature request need to be sent upstream to libtorrent to enable this?

**Comment by eddmanx** [comment]
Yea, they are annoying and user unfriendly. Why does it have to be so many .parts files anyway? uTorrent keeps all that extra data in a single .dat files, meaning it's far more manageable.

People sometimes do not download the entirety of a large torrent and select only a part of it, so the multiple .parts files are just in the way.

**Comment by SpL1Ne** [comment]
Bump.

**Comment by SeriousHoax** [comment]
I have always wanted this option also. It would be nice to have this.
Tixati gives you the option to store them wherever you want which is very nice. I know Tixati is not using libtorrent so it's not the same but still worth mentioning.
<details><summary>Screenshot:</summary>

![Image](https://github.com/user-attachments/assets/12be6422-7ab7-4611-ac5d-85f8ac365e0c)

</details><br/>

**Comment by SeriousHoax** [comment]
Bump

**Comment by SpL1Ne** [comment]
@arvidn bot didn't work properly and still closed issue even though there was futher activity after stale label.

**Comment by robturner45** [comment]
Chiming in with the fact that I find this incredibly annoying also - just give us a single directory option or a per download directory option

Part Files Stored in: 
❎ Path of downloaded file
❎ Subfolder in path of downloaded file
✅ Custom folder for all parts files: \<dir>

**Comment by arvidn** [comment]
maybe something like this: https://github.com/arvidn/libtorrent/pull/8119

**Comment by cheater** [comment]
I haven't read through the diff, but please make sure the parts file path
is relative from the save directory of the torrent, rather than one global
location. Also please make sure I can take old parts files that are
currently cluttering my download directory, put them all in ./.parts/ (or
whatever the subdir would be) using just normal mv, and they will get found
(i.e. the file names are the same)

**Comment by arvidn** [comment]
that patch lets you pick an absolute directory for the part file. it's not relative to the download directory. If the path for the part file is specified, it won't move when the storage for the torrent is moved.

My understanding is that the main problem people have with the way it works by default is that the part files are intermingled with the other downloaded files.

**Comment by glassez** [comment]
> If the path for the part file is specified, it won't move when the storage for the torrent is moved.

This can be a problem in some cases unless you also provide a way to move parts file.

**Comment by arvidn** [comment]
> This can be a problem in some cases unless you also provide a way to move parts file.

I'm open to suggestions on how best to do that

**Comment by glassez** [comment]
> > This can be a problem in some cases unless you also provide a way to move parts file.
> 
> I'm open to suggestions on how best to do that

I can only think of new `torrent_handle::move_part_file` method or additional parameter of `torrent_handle::move_storage`...

**Comment by glassez** [comment]
> that patch lets you pick an absolute directory for the part file. it's not relative to the download directory. 

Why not allow it to be set as relative path to the save path as well? Then it could cover more use cases.

**Comment by HanabishiRecca** [comment]
That sounds like something that could easily be done on the client side.

**Comment by glassez** [comment]
> That sounds like something that could easily be done on the client side.

What exactly?

**Comment by glassez** [comment]
Currently part file location is also can be considered as relative to save path so if you allow it to support other relative paths it will reuse a lot of existing logic.

**Comment by HanabishiRecca** [comment]
> What exactly?

You just update `part_file_path` when `save_path` changes so they are in sync. That way you can implement "relative" feature on the client side.

**Comment by cheater** [comment]
Changing to absolute path only is bad. Please do only relative path, or
allow both. Relative should be default, as it is already - the parts files
are currently relative. The issue is that they're not in a subdir, not that
they're relative, relative is what you actually want.

Bear in mind people can have lots of file systems, torrents can be spread
across them due to space constraints, and those can be taken offline and
online with time (eg laptop + usb drives).

For all users except those with so few torrents that they fit on one file
system, an absolute path will be fragile, it will create problems with
parts disappearing if a specific file system cannot always be brought
online, and can also lead to issues with out of space file systems (because
parts cannot be spread across multiple FSes). Even then, with just one file
system, if the system's drive letter (Windows) or mount point (Linux / ...)
changes, the torrents will break without additional work.

Users will want to be able to plug in one drive and have their torrents
work, having to plug in two drives to get one torrent to work is not
something people should be asked to do.

Regarding moving files:

For torrent downloads which are currently not imported into a torrent
client, the user should be able to move files by hand to the new format (by
simply doing `mv .*.parts .parts/.`) so that:

1. the files stop cluttering the main download directory for now
2. when the torrents are added to a client (qBt) again, and preseeded with
that download directory, the torrent data and parts files are all found
without any issue

If you do the obvious thing this should be exactly how it works, just
wanted to mention this in case you had something special in mind that might
not work like that.
